apps/openmw/mwmechanics/aiwander.cpp

```diff
@@ -206,29 +217,51 @@ namespace MWMechanics
         if (REACTION_INTERVAL <= lastReaction)
         {
             lastReaction = 0;
-            return reactionTimeActions(actor, storage, currentCell, cellChange, pos);
+            return reactionTimeActions(actor, storage, currentCell, cellChange, pos, duration);
         }
         else
             return false;
     }
 
     bool AiWander::reactionTimeActions(const MWWorld::Ptr& actor, AiWanderStorage& storage,
-        const MWWorld::CellStore*& currentCell, bool cellChange, ESM::Position& pos)
+        const MWWorld::CellStore*& currentCell, bool cellChange, ESM::Position& pos, float duration)
     {
+        if (mDistance <= 0)
+            storage.mCanWanderAlongPathGrid = false;
+
         if (isPackageCompleted(actor, storage))
         {
             return true;
         }
 
+        if (!mStoredInitialActorPosition)
+        {
+            mInitialActorPosition = actor.getRefData().getPosition().asVec3();
+            mStoredInitialActorPosition = true;
+        }
+
         // Initialization to discover & store allowed node points for this actor.
         if (mPopulateAvailableNodes)
         {
-            getAllowedNodes(actor, currentCell->getCell());
+            getAllowedNodes(actor, currentCell->getCell(), storage);
         }
 
         // Actor becomes stationary - see above URL's for previous research
-        if(mAllowedNodes.empty())
-            mDistance = 0;
+        // If a creature or an NPC with a wander distance and no pathgrid is available,
+        // randomly idle or wander around near spawn point
+        if(mAllowedNodes.empty() && mDistance > 0 && !storage.mIsWanderingManually) {
+            // Typically want to idle for a short time before the next wander
+            if (Misc::Rng::rollDice(100) >= 96) {
+                wanderNearStart(actor, storage, mDistance);
+            }
+        } else if (mAllowedNodes.empty() && !storage.mIsWanderingManually) {
+            storage.mCanWanderAlongPathGrid = false;
+        }
+
+        // If Wandering manually and hit an obstacle, stop
+        if (storage.mIsWanderingManually && mObstacleCheck.check(actor, duration*10.0f, 2.5f)) {
```

Comment:
The distance was increased to handle it slowly sliding against an obstacle just enough to move, but jittering pretty heavily in the process which looked really bad. Normally for AI_FOLLOW or others, this isn't really a problem since it still is technically making progress towards its destination.